plugins/configs/api/controller.go

```diff
@@ -0,0 +1,917 @@
+	c := &api{
+		config: cfg,
+		agent:  agent,
+		ctx:    ctx,
+	}
+	c.outputCtx, outputCancel = context.WithCancel(context.Background())
+	return c, outputCancel
+}
+
+// PluginConfig is a plugin name and details about the config fields.
+type PluginConfigTypeInfo struct {
+	Name   string                 `json:"name"`
+	Config map[string]FieldConfig `json:"config"`
+}
+
+type PluginConfig struct {
+	ID string `json:"id"` // unique identifer
+	PluginConfigCreate
+}
+
+type PluginConfigCreate struct {
+	Name   string                 `json:"name"`   // name of the plugin
+	Config map[string]interface{} `json:"config"` // map field name to field value
+}
+
+// FieldConfig describes a single field
+type FieldConfig struct {
+	Type      FieldType              `json:"type,omitempty"`       // see FieldType
+	Default   interface{}            `json:"default,omitempty"`    // whatever the default value is
+	Format    string                 `json:"format,omitempty"`     // type-specific format info. eg a url is a string, but has url-formatting rules.
+	Required  bool                   `json:"required,omitempty"`   // this is sort of validation, which I'm not sure belongs here.
+	SubType   FieldType              `json:"sub_type,omitempty"`   // The subtype. map[string]int subtype is int. []string subtype is string.
+	SubFields map[string]FieldConfig `json:"sub_fields,omitempty"` // only for struct/object/FieldConfig types
+}
+
+// FieldType enumerable type. Describes config field type information to external applications
+type FieldType string
+
+// FieldTypes
+const (
+	FieldTypeUnknown     FieldType = ""
+	FieldTypeString      FieldType = "string"
+	FieldTypeInteger     FieldType = "integer"
+	FieldTypeDuration    FieldType = "duration" // a special case of integer
+	FieldTypeSize        FieldType = "size"     // a special case of integer
+	FieldTypeFloat       FieldType = "float"
+	FieldTypeBool        FieldType = "bool"
+	FieldTypeInterface   FieldType = "any"
+	FieldTypeSlice       FieldType = "array"  // array
+	FieldTypeFieldConfig FieldType = "object" // a FieldConfig?
+	FieldTypeMap         FieldType = "map"    // always map[string]FieldConfig ?
+)
+
+// Plugin is an instance of a plugin running with a specific configuration
+type Plugin struct {
+	ID   models.PluginID
+	Name string
+	// State()
+	Config map[string]FieldConfig
+}
+
+func (a *api) ListPluginTypes() []PluginConfigTypeInfo {
+	result := []PluginConfigTypeInfo{}
+	inputNames := []string{}
+	for name := range inputs.Inputs {
+		inputNames = append(inputNames, name)
+	}
+	sort.Strings(inputNames)
+
+	for _, name := range inputNames {
+		creator := inputs.Inputs[name]
+		cfg := PluginConfigTypeInfo{
+			Name:   name,
+			Config: map[string]FieldConfig{},
+		}
+
+		p := creator()
+		getFieldConfig(p, cfg.Config)
+
+		result = append(result, cfg)
+	}
+	// TODO: add more types?
+	return result
+}
+
+func (a *api) ListRunningPlugins() (runningPlugins []Plugin) {
+	if a == nil {
```

Comment:
This func doesn't return whether a plugin is an input or an output. Since we have some plugins with the same name that are inputs and outputs (like inputs.file and outputs.file) we need to disambiguate them here.

Reply:
plugin names should be `inputs.file` or `outputs.file`. There's a few places where this was missed and i'm currently cleaning them up.